tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/TensorScope.java

```diff
@@ -0,0 +1,536 @@
+/*
+  Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ ==============================================================================
+ */
+package org.tensorflow;
+
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+
+/**
+ * A scope that can be used to manage tensor resources.  Any tensors created between a scope's creation and calling
+ * {@code close()} that haven't been detached or attached to a different scope are guaranteed to be closed with the
+ * scope (even if they are created in a sub-scope).  Tensors may be manually closed earlier without issue.
+ * <p>
+ * Tensors are automatically tracked on creation.  A tensor can me manually added to a scope with {@link
+ * TensorScope#attach(Tensor)} or {@link Tensor#attachToCurrent()}.  A tensor may only have one scope: if it currently
+ * has a scope when {@code attach} is called, it is removed from its original scope.
+ * <p>
+ * {@link Tensor#detach()} detaches the tensor from it's scope, requiring the user to close it manually or attach it to
+ * another scope.
+ * <p>
+ * Scopes will be inherited at thread creation, but further scope creation on different threads will be independent,
+ * other than having the same parent. Closing a scope will close it's children regardless of which threads they are on.
+ */
+public class TensorScope implements AutoCloseable {
+
+  private static final InheritableThreadLocal<TensorScope> currentScope = new InheritableThreadLocal<>();
+
+  public static TensorScope currentScope() {
+    TensorScope scope = currentScope.get();
+
+    if (scope == null || !scope.closed) {
+      return scope;
+    }
+
+    // scope could be closed in another thread, in which case this thread's currentScope wouldn't be updated
```

Comment:
I'm wondering, since `Tensor` instances are [not thread-safe](https://github.com/tensorflow/java/blob/b54526c5a75147c9c7c0a3a94b2c08a8e09311b3/tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Tensor.java#L33), should it be the same with tensor scopes?

Reply:
It would make things a bit easier, but if you have something like data loading, you might want one scope to oversee multiple threads, which why this is there.  I don't really see a benefit to removing it, it's not that costly to manage.

Reply:
If tensors cannot be shared across threads, it could be dangerous to group them in a shared container. Even if it is not costly to support it, it might be preferable to be more strict and mention that `TensorScope` are not thread-safe (unless there is a real good use case for supporting it). For example, why would we allow a thread to close resources that it does not "own", according to the non thread-safe constraint of the tensors?

Reply:
As I pointed out in issue #184, simply allowing deterministic deallocation automatically renders everything *not* thread-safe. It's pointless even trying to make it thread-safe.

Reply:
Can tensors not be moved across threads?  The impression I got from "not thread safe" is that accessing the same tensor from different threads at the same time may not work, but that something like the example below (returning tensors across thread boundaries) would work fine.

I'm thinking of something like:
```java
ExecutorService es = ...
TensorScope.withCleanup(() -> {
  Future<Tensor> f1 = es.submit(() -> getTensor(1));
  Future<Tensor> f2 = es.submit(() -> getTensor(2));

  Tensor t1 = f1.get();
  Tensor t2 = f2.get();
});
```

Imo, looking at this, `t1` and `t2` should both be part of the scope, which requires sharing the current scope + hierarchy (which is what is done).

Reply:
@rnett It sounds like what you're after though isn't thread-safety, but reference counting. It's possible to make reference counting thread-safe pretty easily, up until the point where deallocation happens. We can have multiple threads accessing data, and releasing resources as they are done, and the parent thread would be there to do the actual deallocation when everything looks like it's done with reference counts of 1 or something. Check, for example, Netty's ByteBuf and JavaCPP's Pointer.ReferenceCounter to see what that looks like.

Reply:
> Imo, looking at this, `t1` and `t2` should both be part of the scope (and any tensors created in the `getTensor` calls, too, for safety), which requires sharing the current scope + hierarchy (which is what is done).

In general, this simply doesn't work. The getTensor() calls could start looping and attaching stuff on their own, so to prevent running out of memory they may need to release those temporary buffers before returning. At the very least, you're going to need to synchronize all calls to TensorScope.

Reply:
What would this have to do with reference counting?  We're not accessing data from multiple threads, just producing it in workers and then returning it to main.

And yeah, the `getTensor()` calls may need to use TensorScope themselves, internally, but that's fine, nesting works fine even across threads.  `TensorScope` is already synchronized where necessary.

Reply:
How is getTensor() going to access the TensorScope from the main thread?

Reply:
`currentScope`

Reply:
`currentScope` is thread-local storage. It's not going to inherit automagically anything from another thread.

Reply:
Ah, ok, it will, that's what InheritableThreadLocal does... sigh... Well, we could spend all day poking holes at the code to figure out where it fails, but the fact remains that it's not clear at all why someone would want to use all that. It's not obvious that it does anything that one could not do normally with a few List<Tensor> and a couple of calls to close and move those around.

Reply:
My main concern about this is that it is not clear for user what is safe and not safe to do with tensors in a multithread scenario. Right now the statement is pretty clear: Tensors are not thread-safe and the library is not adding any safeguards around it. Here we are adding some support for thread-safety but only for some specific cases that we presume the users might want.

Why not keeping the first version of this feature simple and strict, stating that scopes like tensors cannot be shared across threads, then add support for multithreading later if really needed and maybe even adding it for the whole `Tensor` API?

Reply:
I'm fine with that, it wasn't really clear to me either.  Although if you don't mind I'll leave all the synchronization (with comments) and just remove the `InheritableThreadLocal`.  It won't actually affect anything, and it makes easier for whoever ends up adding it later.